### PR TITLE
blog-api first version

### PR DESCRIPTION
**Pull request by patilsourabh4097**
Description Endpoints

• Home:

### Home page -
URI: GET /
Status: 200
Response: {success: "Connected"}
• Authentication:

### authRoute -
URI: POST/auth/sign-up
Status: 200
Request: {
username,
password,
}
Response: {
    success: "successfully signed up",
    id:newUser._id
  }

URI: POST /auth/log-in
Status: 200
Request: {
username,
password,
}
Response:       {
success: "logged in",
      token
}

### userRoute
URI: GET /user
Status: 200
Response: {
user
}

URI: GET/user/:name
Status: 200
Response: {
    username: user.userName,
    posts: user.posts,
  }

### postRoute

URI: GET /post/:posttId
Status: 200
Response: {
    success: "Success",
    data: post,
  }

URI: GET /post/
Status: 200
Response: get all posts

URI: POST /post/
Status: 200
Request: {
title: 
desc,
userId,
}
Response: {
    success: "Post Added",
  }

### commentRoute

URI: POST/post/:postId/comment
Status: 200
req.body = { description }
Response: {
    success: "Comment added",
  }

URI: GET/post/:postId/comments
Status: 200
Response: All comments on postId

### Database schema 
`user` :
{
  userName: String,
  password: String,
  posts: [{ type: mongoose.Schema.Types.ObjectId, ref: "Posts" }]
}

`posts` :
{
  title: String,
  description: String,
  comments: [{ type: mongoose.Schema.Types.ObjectId, ref: "Comment" }],
  userId: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
  date: { type: Date, default: Date.now },
}
`comments` :
{
  description: String,
  date: { type: Date, default: Date.now },
  userId: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
  posts: { type: mongoose.Schema.Types.ObjectId, ref: "Posts" },
}
   

Link of the website - https://git.heroku.com/blogger-api-odin.git

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+const cors = require("cors");
+let dotEnv = require("dotenv");
```
Comment (coder1033) [inline app.js]: `const`

File: app.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+const express = require("express");
+const mongoose = require("mongoose");
+const PORT = process.env.PORT || 5000;
```
Comment (coder1033) [inline app.js]: Put this after all the require statements.

File: app.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+const app = express();
+app.use(cors());
+app.use(express.json());
+app.use(express.urlencoded({ extended: true }));
+
+app.use("/api/auth", authRoute);
+app.use("/api/post", commentRoute);
+app.use("/api/post", postsRoute);
+app.use("/api/user", userRoute);
```
Comment (coder1033) [inline app.js]: Remove the prefix `/api` from every route.

File: app.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+dotEnv.config();
+
+let MONGODB = process.env.MONGODB;
```
Comment (coder1033) [inline app.js]: `const`

File: controller/auth-controller.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+  const user = await User.findOne({ userName: name });
+  if (user) {
+    res.json({
+      err: "User already exists",
+    });
+    return;
+  }
+
+  hashedPassword = await bcrypt.hash(password, 10);
```
Comment (coder1033) [inline controller/auth-controller.js]: `const hashedPassword`

File: controller/auth-controller.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+  let newUser = new User({
+    userName: name,
+    password: hashedPassword,
+  });
+
+  await newUser.save((err) => {
```
Comment (coder1033) [inline controller/auth-controller.js]: I think you've missed to reassign to `newUser`.

File: controller/comment-controller.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  const { postId } = req.params;
+  const isValid = validation.isValidObjectId(postId);
+  if (!isValid) {
+    res.json({
+      err: "Invalid Id",
+    });
+    return;
+  }
+  const { description } = req.body;
+  const userId = req.user._id;
```
Comment (coder1033) [inline controller/comment-controller.js]: You can make these `req` assignments at the top of the function body.

File: controller/posts-controller.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
+const Comment = require("../models/comment");
+const Posts = require("../models/posts");
+const User = require("../models/user");
+
+const validation = require("./validation");
+
+exports.addPost = async (req, res) => {
+  const { title, description } = req.body;
+  const userId = req.user._id;
+  let newPost = new Posts({
```
Comment (coder1033) [inline controller/posts-controller.js]: Instead of  `newPost` you can just use `post` here.

File: controller/posts-controller.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    title,
+    description,
+    userId,
+  });
+
+  let post = await newPost.save();
+  postId = post._id;
```
Comment (coder1033) [inline controller/posts-controller.js]: `const postId`

File: controller/posts-controller.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+  let post = Posts.findById(postId).populate("userid", "userName");
+  post.populate({
+    path: "Comments",
+    populate: {
+      path: "User",
+      select: "userName",
+    },
+  });
+  post = await post;
```
Comment (coder1033) [inline controller/posts-controller.js]: What's this?

File: controller/user-controller.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
+const User = require("../models/user");
+
+exports.getCurrentUser = async (req, res) => {
+  const { userName } = req.user;
+  const userId = req.user._id;
+  let user = {
```
Comment (coder1033) [inline controller/user-controller.js]: `const`

File: controller/validation.js [hunk 12]
```diff
@@ -0,0 +1,6 @@
+const mongoose = require("mongoose");
+
+exports.isValidObjectId = (id) => {
+  let isValid = mongoose.Types.ObjectId.isValid(id);
+  return isValid;
+};
```
Comment (coder1033) [inline controller/validation.js]: This should not be in the controllers directory.
Make a separate directory for validators.

File: models/comment.js [hunk 13]
```diff
@@ -0,0 +1,10 @@
+const mongoose = require("mongoose");
+
+let Comment = new mongoose.Schema({
```
Comment (coder1033) [inline models/comment.js]: `const`

File: models/comment.js [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+  description: String,
+  date: { type: Date, default: Date.now },
+  userId: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
```
Comment (coder1033) [inline models/comment.js]: `user` instead of `userId`

File: models/comment.js [hunk 15]
```diff
@@ -0,0 +1,10 @@
...
+  posts: { type: mongoose.Schema.Types.ObjectId, ref: "Posts" },
```
Comment (coder1033) [inline models/comment.js]: `post` instead of `posts`

File: models/posts.js [hunk 16]
```diff
@@ -0,0 +1,11 @@
+const mongoose = require("mongoose");
+
+let Posts = new mongoose.Schema({
+  title: String,
+  description: String,
+  comments: [{ type: mongoose.Schema.Types.ObjectId, ref: "Comment" }],
+  userId: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
```
Comment (coder1033) [inline models/posts.js]: `user` instead of `userId`

File: models/posts.js [hunk 17]
```diff
@@ -0,0 +1,11 @@
+const mongoose = require("mongoose");
+
+let Posts = new mongoose.Schema({
```
Comment (coder1033) [inline models/posts.js]: `const Post`

File: models/posts.js [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+  title: String,
+  description: String,
+  comments: [{ type: mongoose.Schema.Types.ObjectId, ref: "Comment" }],
+  userId: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
+  date: { type: Date, default: Date.now },
+});
+
+module.exports = mongoose.model("Posts", Posts);
```
Comment (coder1033) [inline models/posts.js]: `Post` instead of `Posts`.

Also this file name should be `post.js`.

**Comment by coder1033** [comment]
How you're getting all the comments of a particular post?

File: app.js (hunk 2 above)
Comment (patilsourabh4097) [reply app.js]: what, PORT?

File: controller/auth-controller.js (hunk 5 above)
Comment (patilsourabh4097) [reply controller/auth-controller.js]: `updated`

File: controller/auth-controller.js (hunk 6 above)
Comment (patilsourabh4097) [reply controller/auth-controller.js]: done

File: controller/comment-controller.js (hunk 7 above)
Comment (patilsourabh4097) [reply controller/comment-controller.js]: updated

File: controller/posts-controller.js (hunk 8 above)
Comment (sourabh-4097) [reply controller/posts-controller.js]: updated

File: controller/posts-controller.js (hunk 10 above)
Comment (sourabh-4097) [reply controller/posts-controller.js]: updated

File: models/posts.js (hunk 18 above)
Comment (patilsourabh4097) [reply models/posts.js]: updated

File: controller/validation.js (hunk 12 above)
Comment (patilsourabh4097) [reply controller/validation.js]: updated

File: controller/posts-controller.js [hunk 19]
```diff
@@ -1,24 +1,24 @@
...
+  let post = new Post({
     title,
     description,
-    userId,
+    user:userId,
   });
 
-  let post = await newPost.save();
-  postId = post._id;
+  let post = await post.save();
```
Comment (coder1033) [inline controller/posts-controller.js]: You've already declared `post` in this function. You should remove `let` here.

File: controller/posts-controller.js [hunk 20]
```diff
@@ -55,15 +55,14 @@ exports.getSinglePost = async (req, res) => {
     return;
   }
 
-  let post = Posts.findById(postId).populate("userid", "userName");
+  let post = await Post.findById(postId).populate("user", "userName");
```
Comment (coder1033) [inline controller/posts-controller.js]: Again, why you're using `let` here?

Comment (patilsourabh4097) [reply controller/posts-controller.js]: I am using `post` again in code..

File: controller/posts-controller.js (hunk 19 above)
Comment (patilsourabh4097) [reply controller/posts-controller.js]: updated

File: routes/post-route.js [hunk 21]
```diff
@@ -0,0 +1,12 @@
+const express = require("express");
+
+const authController = require("../controller/auth-controller");
+const postsController = require("../controller/posts-controller");
+
+const router = express.Router();
+
+router.post("/", authController.isAuth, postsController.addPost);
+router.get("/:postId", postsController.getSinglePost);
+router.get("/user/:name", postsController.getUserPost);
```
Comment (coder1033) [inline routes/post-route.js]: Full URI - `/user/:username/posts`

**Comment by coder1033** [comment]
Also make sure to update PR description.

File: app.js [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+mongoose.connect(MONGODB, {
+  useNewUrlParser: true,
+  useUnifiedTopology: true,
+  useFindAndModify: false,
+});
+
+app.get("/", (req, res) => {
+  res.json({
+    success: "Connected blog-api",
+  });
```
Comment (coder1033) [inline app.js:35]: What is the response status here?

File: controller/auth-controller.js [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+exports.login = async (req, res) => {
+  const {name, password} = req.body
+  const key = process.env.JWT_SECRET;
+
+  const user = await User.findOne({ userName: name });
+
+  if (!user) {
+    res.json({
+      err: "User dosent exist",
+    });
```
Comment (coder1033) [inline controller/auth-controller.js:15]: What is the Response status here?
Please look into each response.

File: controller/auth-controller.js [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+
+exports.isAuth = (req, res, next) => {
+  if (!req.headers.auth) {
+    res.json({
+      err: "not logged in",
+    });
+    return;
+  }
+
+  let token = req.headers.auth;
```
Comment (coder1033) [inline controller/auth-controller.js]: Assign this at the top of function body?

File: controller/comment-controller.js [hunk 25]
```diff
@@ -0,0 +1,48 @@
+const Comment = require("../models/comment");
+const Post = require("../models/post");
+
+const validation = require("../validators/validation");
+
+exports.addComment = async (req, res) => {
+  const { postId } = req.params;
+  const { description } = req.body;
+  const user = req.user._id;
```
Comment (coder1033) [inline controller/comment-controller.js]: `userId`
Please be consistent on naming.

Comment (coder1033) [reply controller/comment-controller.js]: Make this change everywhere.

File: controller/posts-controller.js (hunk 20 above)
Comment (coder1033) [reply controller/posts-controller.js]: You can combine the next populate expression to this expression itself.

File: controller/user-controller.js [hunk 26]
```diff
@@ -0,0 +1,32 @@
+const User = require("../models/user");
+
+exports.getCurrentUser = async (req, res) => {
```
Comment (coder1033) [inline controller/user-controller.js]: Can you explain what exactly you want from this current user?

File: controller/user-controller.js [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+      username: userName,
+      id: userId,
+    },
+  };
+  res.json(user);
+};
+
+exports.getUser = async (req, res) => {
+  const { name } = req.params;
+  const user = await User.findOne({ userName: name });
```
Comment (coder1033) [inline controller/user-controller.js]: You might want to populate the `posts`.

File: controller/comment-controller.js (hunk 25 above)
Comment (patilsourabh4097) [reply controller/comment-controller.js]: earlier it was userId, but you said to change it to user, hence changed..

File: controller/user-controller.js (hunk 26 above)
Comment (patilsourabh4097) [reply controller/user-controller.js]: the user who is logged in currently..

File: controller/posts-controller.js (hunk 20 above)
Comment (patilsourabh4097) [reply controller/posts-controller.js]: updated

File: app.js (hunk 22 above)
Comment (patilsourabh4097) [reply app.js:35]: updated

File: controller/auth-controller.js (hunk 23 above)
Comment (patilsourabh4097) [reply controller/auth-controller.js:15]: updated

File: controller/auth-controller.js (hunk 24 above)
Comment (patilsourabh4097) [reply controller/auth-controller.js]: done

File: controller/user-controller.js (hunk 27 above)
Comment (patilsourabh4097) [reply controller/user-controller.js]: updated

File: app.js [hunk 28]
```diff
@@ -0,0 +1,40 @@
...
+const PORT = process.env.PORT || 5000;
+
+const app = express();
+app.use(cors());
+app.use(express.json());
+app.use(express.urlencoded({ extended: true }));
+
+app.use("/auth", authRoute);
+app.use("/post", commentRoute);
+app.use("/post", postsRoute);
```
Comment (coder1033) [inline app.js]: `postRoute`

**Review by coder1033 — APPROVED** [review]

**Pull request merged**